lib/internal/native_module.js

```diff
@@ -0,0 +1,217 @@
+// This file creates the internal module & binding loaders used by built-in
+// modules. In contrast, user land modules are loaded using
+// lib/internal/module.js (CommonJS Modules) or
+// lib/internal/loader/* (ES Modules).
+//
+// This file is compiled an run by node.cc before bootstrap_node.js
+// was called, therefore the loaders are bootstraped before we start to
+// actually bootstrap Node.js. It creates the following objects:
+//
+// C++ binding loaders:
+// - process.binding: the legacy C++ binding loader, accessible from user land
+//   because it is an object attached to the global process object.
+//   These C++ bindings are created using NODE_BUILTIN_MODULE_CONTEXT_AWARE().
+//   We do not make any guarantees about the stability of these bindings,
+//   but still have to take care of compatibility issues caused by them
+//   from time to time.
+// - internalBinding: the private internal C++ binding loader, inaccessible
+//   from user land because it is directly passed to NativeModule's wrapper.
+//   These C++ bindings are created using NODE_MODULE_CONTEXT_AWARE_INTERNAL().
+//
+// Internal JavaScript module loader:
+// - NativeModule: a minimal module system used to load the JavaScript core
+//   modules found in lib/**/*.js and deps/**/*.js. All core modules are
+//   compiled into the node binary via node_javascript.cc generated by js2c.py,
+//   so they can be loaded faster without the cost of I/O. This class makes the
+//   lib/internal/*, deps/internal/* modules and internalBinding() available by
+//   default to core modules, and lets the core modules require itself via
+//   require('internal/native_module') even when this file is not written in
+//   CommonJS style.
+//
+// Other objects:
+// - process.moduleLoadList: an array recording the bindings and the modules
+//   loaded in the process and the order in which they are loaded.
+
+'use strict';
+
+(function bootstrapInternalLoaders(process) {
+  const moduleLoadList = [];
+  Object.defineProperty(process, 'moduleLoadList', {
+    value: moduleLoadList,
+    configurable: true,
+    enumerable: true,
+    writable: false
+  });
+
+  // Set up process.binding
+  {
+    const bindingObj = Object.create(null);
+
+    const getBinding = process.binding;
+    process.binding = function binding(module) {
+      module = String(module);
+      let mod = bindingObj[module];
+      if (typeof mod !== 'object') {
+        mod = bindingObj[module] = getBinding(module);
+        moduleLoadList.push(`Binding ${module}`);
+      }
+      return mod;
+    };
+
+    const getLinkedBinding = process._linkedBinding;
+    process._linkedBinding = function _linkedBinding(module) {
+      module = String(module);
+      let mod = bindingObj[module];
+      if (typeof mod !== 'object')
+        mod = bindingObj[module] = getLinkedBinding(module);
+      return mod;
+    };
+  }
+
+  // Set up internalBinding in the closure
```

Comment:
Come to think of it, we may also put `internalBinding` in the returned object and rename this file `inernal_loaders` which seem to be more accurate..so it will become `const { NativeModule, internalBinding } = require('internal/internal_loaders')` (somewhat clashes with `internal/loader/*` though..)